cloudinit/osys/network.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2015 Cloudbase Solutions Srl
+# Copyright 2015 Canonical Ltd.
+# License TBD
+
+import abc
+import collections
+
+import six
+
+from cloudinit.osys import util
+
+
+__all__ = (
+    'BaseNetwork',
+    'BaseRoute',
+    'BaseRouteCollection',
+    'BaseInterface',
+)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseNetwork(object):
+    """Base network class for network related utilities."""
+
+    @abc.abstractmethod
+    def routes(self):
+        """Get an object representing the available routes."""
+
+    @abc.abstractmethod
+    def default_gateway(self):
+        """Get the default gateway, as a route object."""
+
+    @abc.abstractmethod
+    def interfaces(self):
+        """Get an object representing the available interfaces."""
+
+    @abc.abstractmethod
+    def hosts(self):
+        """Get an object representing the available hosts."""
+
+    @abc.abstractmethod
+    def set_hostname(self, hostname):
+        """Change the host name of the instance."""
+
+    @abc.abstractmethod
+    def set_timezone(self, timezone):
+        """Change the current timezone with the given timezone."""
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseRoute(object):
+    """Base class for routes."""
+
+    def __init__(self, destination, gateway, netmask,
+                 interface, metric,
+                 flags=None, refs=None, use=None, expire=None):
+        self.destination = destination
+        self.gateway = gateway
+        self.netmask = netmask
+        self.interface = interface
+        self.metric = metric
+        self.flags = flags
+        self.refs = refs
+        self.use = use
+        self.expire = expire
+
+    @abc.abstractproperty
+    def static(self):
+        """Check if this route is static."""
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BaseRouteCollection(collections.Sequence):
+    """A collection which encapsulates the routes available in the system.
+
+    Each route item will be an instance of :class:`BaseRoute` and a particular
+    implementation depending on the underlying platform.
+    To retrieve all the routes, it is sufficient to instantiate this class,
+    it will, under the hood, retrieve the routes.
+    The collection offers a couple of methods for adding or deleting a route.
+
+    >>> RouteCollection.add(Route(..., ...))
+    >>> RouteCollection.delete(Route(..., ...))
+    >>> list(RouteCollection())
+    >>> Route(..., ...) in RouteCollection()
+    """
+
+    def __init__(self):
+        self._route_items = self._routes()
```

Comment:
So this feels sorta odd; but maybe its just me. Why not just lazily populate this information instead of populating it before the object is fully constructed?